### ISSUE TITLE
Importing spacy (or thinc) breaks dot product

### ISSUE THREAD

**Issue by jona-sassenhagen** [explosion/spaCy#12807]
A very annoying bug that it took me forever to track down ... I'm at a loss for what might be going on here.

## How to reproduce the behaviour
```
import numpy as np

# works
small = np.random.randn(5, 6)
small.T @ small

# works
larger = np.random.randn(100, 110)
larger.T @ larger

import spacy  # or import thinc

# works
small.T @ small

# works
larger + larger

# hangs forever, max CPU usage
larger.T @ larger  # larger.T.dot(larger) also hangs
```

## Your Environment
- **spaCy version:** 3.5.3 (also occurs under 3.6.0)
- **Platform:** macOS-13.4.1-x86_64-i386-64bit
- **Python version:** 3.11.4
- **Pipelines:** de_core_news_md (3.5.0), en_core_web_trf (3.5.0), en_core_web_sm (3.5.0), en_core_web_md (3.5.0)

I tried the same thing on another computer (Raspberry Pi) and it worked flawlessly, but on my Macbook, it hangs.

I can move this to the thinc github if you prefer.

**Comment by MohamedTaha314** [comment]
Same here! 
Spacy version 3.6.0 and Python version 3.9.5

Apparently it has something to do with pydantic and typing_extensions libraries versions, creating a conflict

Does anybody have a fix?

**Comment by svlandeg** [comment]
@jona-sassenhagen: thanks for the report, and I'm sorry you've been having issues with this!

I've been trying to replicate the issue you're seeing, but on my system the commands just execute without delay or hanging. I don't have a Macbook though.

Just for my understanding: the same code runs without issue if you simply remove the `import spacy` statement? And if you do `import`, the first two statements run fine and then the 3rd and last hangs? And this happens consistently, but only on your Macbook?

**Comment by svlandeg** [comment]
@MohamedTaha314: are you certain this is the same problem? Are you running into the same problem that the above code executes fine without the `import spacy` statement, but hangs if you add the import?

If it's not exactly the same, I suggest opening a new issue instead.

If it's the exact same behaviour as described, can you elaborate on why it'd have something to do with pydantic and typing_extensions libraries versions? Is there some kind of warning about a conflict, and if so, can you share that in full? (please share this in the new issue if the problem is different than the original post, though)

**Comment by jona-sassenhagen** [comment]
@svlandeg No need to apologize, you provide amazing software for free :) And I have a hunch this issue is very idiosyncratic - it is unlikely to affect everyone or even a large fraction of those using spacy on a Mac otherwise there'd be a lot more complaints.

Anyways to clarify:

This works:
```import numpy as np
larger = np.random.randn(100, 110)
larger.T @ larger```

This also terminates succesfully:
```import numpy as np
import spacy

small = np.random.randn(5, 6)
small.T @ small```

This however takes up all CPU forever:
```import numpy as np
import spacy

larger = np.random.randn(100, 110)
larger.T @ larger```

I noticed this when trying to run `sklearn.metrics.pairwise.cosine_similarity` on some spacy vectors.

There may be some additional conditions I haven't found yet.

**Comment by jona-sassenhagen** [comment]
Oh and this also hangs:

```
import numpy as np
import thinc

larger = np.random.randn(100, 110)
larger.T @ larger
```

I can't even ctrl-c the process.

**Comment by svlandeg** [comment]
Huh, I'm pretty stumped by this issue. A colleague of mine just tested on an M1 Macbook Mini and couldn't reproduce the hanging - he just got the output straight away.

This is an annoying suggestion, but have you tried in an entirely clean `venv`? I just wonder whether something else is somehow interfering.

**Comment by jona-sassenhagen** [comment]
I'm on an Intel Mac. I doubt it affects Apple Silicon.

>Like - I don't suppose you have any custom functions that you're adding to the registry or something such, that get loaded when spacy or thing are imported?

Not to my knowledge, but I'll test a new env.

I actually suspect it'll just work there to be honest. It's just annoying that it doesn't work in my base environment ...

**Comment by jona-sassenhagen** [comment]
Ok I set up a new clean conda env with the default python (3.11), conda install pip, pip install numpy spacy.

Works fine. I guess unless anyone has a good idea, I'll have to close as irreproducible.

**Comment by svlandeg** [comment]
It's definitely annoying that it doesn't work in your base environment, yea. You could always try building up your new `venv` further until the bug hits again :trollface:

**Comment by jona-sassenhagen** [comment]
It's fine, thanks.

**Comment by jona-sassenhagen** [comment]
I can confirm `pip install typing_extensions==4.2.0` fixed the problem for me!

**Comment by svlandeg** [comment]
Oh wow, thanks both for digging further into this! That's definitely surprising to me.

**Comment by jona-sassenhagen** [comment]
It's good to have a workaround, but I'm curious - why would this happen?

**Comment by svlandeg** [comment]
We looked into this more internally. It does feel like this is coming from an external library, but none of us can reproduce this and we haven't been able to find related bug reports in the libraries that spaCy depends on. So for now I'm afraid there's not much else we can do except recommend users to `pip install typing_extensions==4.2.0` when this happens...

I'll go ahead and close out this issue as there's not much else to do on our end. Thanks again for the report though, definitely appreciate it.

**Comment by jona-sassenhagen** [comment]
Sure makes sense, thanks!
